src/hotspot/share/opto/vmaskloop.cpp

```diff
@@ -0,0 +1,975 @@
+/*
+ * Copyright (c) 2023, Arm Limited. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "precompiled.hpp"
+#include "opto/addnode.hpp"
+#include "opto/convertnode.hpp"
+#include "opto/vmaskloop.hpp"
+
+//        L O O P   V E C T O R   M A S K   T R A N S F O R M A T I O N
+// ============================================================================
+
+// -------------------------------- Constructor -------------------------------
+VectorMaskedLoop::VectorMaskedLoop(PhaseIdealLoop* phase) :
+  _phase(phase),
+  _igvn(&(phase->_igvn)),
+  _arena(phase->C->comp_arena()),
+
+  _lpt(nullptr),
+  _cl(nullptr),
+  _cle(nullptr),
+  _iv(nullptr),
+
+  _core_set(_arena),
+  _body_set(_arena),
+  _body_nodes(_arena, 32, 0, nullptr),
+  _rpo_idx(_arena, 32, 0, 0),
+  _elem_bt(_arena, 32, 0, T_ILLEGAL),
+  _stmts(_arena, 2, 0, nullptr),
+  _swptrs(_arena, 8, 0, nullptr),
+  _size_stats(_arena)
+{}
+
+// ------------------- Entry function of vector masked loop -------------------
+void VectorMaskedLoop::try_vectorize_loop(IdealLoopTree* lpt) {
+  assert(UseMaskedLoop, "Option should be enabled");
+  assert(lpt->is_counted(), "Loop must be counted");
+  assert(lpt->is_innermost(), "Loop must be innermost");
+
+  CountedLoopNode *cl = lpt->_head->as_CountedLoop();
+  // Skip if loop is already vector masked
+  if (cl->is_vector_masked()) return;
+  // Skip non-post loop
+  if (!cl->is_post_loop()) return;
+  // Skip malformed counted loop
+  if (!cl->is_valid_counted_loop(T_INT)) return;
+  // Skip loop if stride is unsupported
+  if (abs(cl->stride_con()) != 1) return;
+  // Skip loop with control flow
+  if (cl->loopexit()->in(0) != cl) return;
+  // Skip if some loop operations are pinned to the backedge
+  if (cl->back_control()->outcnt() != 1) return;
```

Comment:
It would be interesting to have some trace flag that tells us why we bailed out here and did not do the post-loop vectorization.

Reply:
Great suggestion! Done.

Reply:
Thanks :)